File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    void helper(vector<int>& candidates, vector<int>& ds, int index, int sum,
                int target, vector<vector<int>>& ans) {

        // base...
        if (index == candidates.size()) {
            if (sum == target) {
                ans.push_back(ds);
            }
            return;
        }
        // ...
        if (sum == target) {
            ans.push_back(ds);
            return;
        }

        if (sum < target) {
            ds.push_back(candidates[index]);
            helper(candidates, ds, index, sum + candidates[index], target, ans);
            ds.pop_back();
        }
        helper(candidates, ds, index + 1, sum, target, ans);
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {

        vector<vector<int>> ans;
        vector<int> ds = {};
        helper(candidates, ds, 0, 0, target, ans);
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {

    void helper(int[] candidates, int target, int index, List<Integer> ds, List<List<Integer>> ans) {
        if (target == 0) {
            ans.add(new ArrayList<>(ds));
            return;
        }
        if (index == candidates.length || target < 0) {
            return;
        }
        // pick..
        ds.add(candidates[index]);
        helper(candidates, target - candidates[index], index, ds, ans);
        ds.removeLast();
        helper(candidates, target, index + 1, ds, ans);

    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<Integer> ds = new ArrayList<>();
        List<List<Integer>> ans = new ArrayList<>();
        helper(candidates, target, 0, ds, ans);
        return ans;
    }
}